### ISSUE TITLE
Question: Can ImmutableAttribute be used on return values?

### ISSUE THREAD

**Issue by bill-poole** [dotnet/orleans#8441]
The [documentation](https://learn.microsoft.com/en-us/dotnet/orleans/host/configuration-guide/serialization-immutability) states that "method signatures can include `ImmutableAttribute` on a per-parameter basis." But what about return values?

``` csharp
public interface ISummerGrain : IGrain
{
  // `values` will not be copied. But what about the return value?
  [return: Immutable]
  ValueTask<List<int>> Sum([Immutable] List<int> values);
}
```

Note also that the [documentation](https://learn.microsoft.com/en-us/dotnet/api/orleans.concurrency.immutableattribute?view=orleans-3.0&viewFallbackFrom=orleans-7.0) for `ImmutableAttribute` is only available for v3.6.0. There is nothing available for v7.x.

After downloading the NuGet package, I was able to see that `ImmutableAttribute` is has the `ReturnValue` attribute target, meaning that the above is syntactically valid. But I'd like to confirm that the Orleans runtime actually recognizes and respects `ImmutableAttribute` applied to a grain method return value. If so, it would also be great if the documentation were to state that explicitly.

**Comment by ReubenBond** [comment]
No, not yet. It has that enum value defined but it's not yet implemented.

**Comment by bill-poole** [comment]
Will applying [Immutable] to a Protobuf-serialized type work?

**Comment by ReubenBond** [comment]
That would have to be a feature of the Protobuf codec, but that is not currently implemented. When Orleans delegates to a codec, it does so completely.

**Comment by bill-poole** [comment]
Would applying [Immutable] to a parameter of a grain method work if that parameter is Protobuf-serialized? If so, then we have the ability to specify a Protobuf-serialized parameter as immutable with an [Immutable] attribute applied to the method parameter, but the only way to do so for a return value is to return `Immutable<T>`. Is that correct?

**Comment by ReubenBond** [comment]
Actually, yes, because we generate a closure to represent each interface method, so the immutable attribute on the parameter is actually instructing Orleans that the field on that closure is immutable.

**Comment by bill-poole** [comment]
Okay, so presumably that would apply also to return values if Orleans were to be updated to respect an [Immutable] attribute applied to a return value. Shall I open a separate issue to cover that?
